storey/steps/sample.py

```diff
@@ -42,31 +49,59 @@ def __init__(
         self._window_size = window_size
         self._emit_period = emit_period
         self._emit_before_termination = emit_before_termination
-
+        self._per_key_count = dict()
         self._count = 0
         self._last_event = None
+        self._extract_key: Callable[[Event], str] = self._create_key_extractor(key)
+
+    @staticmethod
+    def _create_key_extractor(key) -> Callable:
+        if key is None:
+            return lambda event: None
+        elif callable(key):
+            return key
+        elif isinstance(key, str):
+            if key == '$key':
+                return lambda event: event.key
+            else:
+                return lambda event: event.body[key]
+        else:
+            raise ValueError(f'Unsupported key type {type(key)}')
 
     async def _do(self, event):
         if event is _termination_obj:
             if self._last_event is not None:
                 await self._do_downstream(self._last_event)
             return await self._do_downstream(_termination_obj)
         else:
-            self._count += 1
+            key = self._extract_key(event)
+
+            if key is not None:
```

Comment:
I'm pretty sure you could remove these checks and `self._count` and everything would still work. Something to consider.